### DISCUSSION TITLE
Trying to add ODIC to my config

### DISCUSSION THREAD

**Discussion by Livefour2day** [authelia/authelia#7057] [Troubleshooting] [answered]
Hi forgive me I am learning, I have been running Authelia for years but recently wanted to add ODIC have generated the certs etc and but I am getting this error and wondered if anyone could point me in the right direction.

level=error msg="Configuration: secrets: error loading secret into key 'identity_providers.oidc.issuer_private_key': it's already defined in other configuration sources" stack="github.com/authelia/authelia/v4/internal/commands/context.go:324

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

The error message you're encountering indicates that the `identity_providers.oidc.issuer_private_key` configuration key has been defined in more than one place within your Authelia configuration. This could happen if you've specified the key in both the configuration file and as an environment variable, or if it's been defined in multiple configuration files that are being loaded by Authelia.

To resolve this issue, you should:

1. Check your Authelia configuration file(s) to locate where `identity_providers.oidc.issuer_private_key` is defined. Ensure it is only defined once in the configuration files.
2. If you are using environment variables to configure Authelia (which is common for sensitive values like keys), ensure that you haven't accidentally exported the `identity_providers.oidc.issuer_private_key` or its corresponding environment variable (`AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE`) more than once.
3. Remember that as of Authelia version 4.38.0, the use of `identity_providers.oidc.issuer_private_key` is deprecated and replaced by `identity_providers.oidc.jwks`. Although the system automatically maps the old configuration to the new one, it's recommended to update your configuration to use `jwks` to avoid future issues and to eliminate deprecation warnings. This change was noted to help users transition smoothly without immediate action but planning for this update is advisable.

After making the necessary adjustments, restart Authelia to apply the changes. If you continue to face issues, double-check the paths and ensure that the environment variables (if used) are correctly set up and not duplicated in any startup scripts or service definitions.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
Looks like you've configured it in multiple locations like secrets, env, and the config file. Can you show your compose envs, and the oidc config section? You can redact the content of the secrets as long as the context exists, see an acceptable example here of a redacted private key:

https://www.authelia.com/configuration/identity-providers/openid-connect/provider/#jwks

**Comment by Livefour2day** [comment]
Hi James, thank you for coming back to me here is my config etc -
ENV -
KEY - AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE
Value - /config/.secrets/certificates/private.pem

OIDC Conig (Redacted) -

identity_providers:
  oidc:
    ## The other portions of the mandatory OpenID Connect 1.0 configuration go here.
    ## See: https://www.authelia.com/c/oidc
    ##### generate with 'docker run authelia/authelia:latest authelia crypto rand --length 128 --charset alphanumeric'
    hmac_secret: .......................
    issuer_private_key: |
      -----BEGIN RSA PRIVATE KEY-----
      .............................
      -----END RSA PRIVATE KEY-----
    clients:
    - id: immich
      description: Immich
      # generate with 'docker run authelia/authelia:latest authelia crypto hash generate pbkdf2 --variant sha512 --random --random.length 72 --random.charset rfc3986'
      secret: .............................
      public: false
      authorization_policy: two_factor
      redirect_uris:
        - https://removed.actual.address/auth/login
      scopes:
        - openid
        - profile
        - groups
        - email
      userinfo_signing_algorithm: none

Also have this at the beginning of my Authelia config but I have Hashed this out to try to eliminate it -
certificates_directory: /config/.secrets/certificates

Sorry James still learning :-)

**Comment by james-d-elliott** [comment, marked as answer]
Yeah you need to either remove it from the YAML config, or remove `AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE` env var.

**Reply by Livefour2day** [reply]
Thanks James,
I removed it from my Config but the error remained so I added it back and removed the ENV, but now I get this error -

level=error msg="Configuration: error occurred during unmarshalling configuration: 1 error(s) decoding:\n\n* error decoding 'identity_providers.oidc.issuer_private_key': could not decode to a *rsa.PrivateKey: x509: failed to parse private key (use ParsePKCS8PrivateKey instead for this key format)" stack="github.com/authelia/authelia/v4/internal/commands/context.go:324 (*CmdCtx).ConfigValidateLogRunE\ngithub.com/authelia/authelia/v4/internal/commands/context.go:186 NewRootCmd.(*CmdCtx).ChainRunE.func1\ngithub.com/spf13/cobra@v1.8.0/command.go:968

**Reply by james-d-elliott** [reply]
How did you generate the key?

**Reply by Livefour2day** [reply]
I used Authlia Console with the following script and changed the path to my .secrets folder
Authelia crypto pair rsa generate --directory /path/to/keys

When I was getting fatal errors I reverted back to my working Authelia until I could figure it out and I renamed the certs folder so I could continue with Authelia which my original config had no certs set up so I renamed the folder to prevent Authelia from getting errors.

**Reply by james-d-elliott** [reply]
Odd. What version are you using to generate it and what version are you running the config with?

**Reply by Livefour2day** [reply]
both are v4.38.7

When I was getting fatal errors I reverted back to my working Authelia until I could figure it out and I renamed the certs folder so I could continue with Authelia which my original config had no certs set up so I renamed the folder to prevent Authelia from getting errors.
Would renaming folders and configs be causing this?

**Reply by james-d-elliott** [reply]
Why does it have `-----BEGIN RSA PRIVATE KEY-----` then? Latest version generates PKCS8 keys with the  `-----BEGIN PRIVATE KEY-----` PEM block header. Also folder wont affect it.

**Reply by Livefour2day** [reply]
well spotted!!! 
I copied the Authelia example config from the web but my Cert does have -----BEGIN PRIVATE KEY----- I didn't copy this to the config :-(

**Reply by james-d-elliott** [reply]
Yeah makes sense. We use the old PKCS1 format to clearly indicate it must be an RSA key. Maybe we should change that though.

**Reply by Livefour2day** [reply]
Getting this error....

time="2024-04-01T11:07:20+01:00" level=error msg="could not import certificate private.pem" stack="github.com/spf13/cobra@v1.8.0/command.go:983            (*Command).execute\ngithub.com/spf13/cobra@v1.8.0/command.go:1115           (*Command).ExecuteC\ngithub.com/spf13/cobra@v1.8.0/command.go:1039           (*Command).Execute\ngithub.com/authelia/authelia/v4/cmd/authelia/main.go:10 main\nruntime/internal/atomic/types.go:194                    (*Uint32).Load\nruntime/asm_amd64.s:1695                                goexit"
time="2024-04-01T11:07:20+01:00" level=error msg="could not import certificate public.pem" stack="github.com/spf13/cobra@v1.8.0/command.go:983            (*Command).execute\ngithub.com/spf13/cobra@v1.8.0/command.go:1115           (*Command).ExecuteC\ngithub.com/spf13/cobra@v1.8.0/command.go:1039           (*Command).Execute\ngithub.com/authelia/authelia/v4/cmd/authelia/main.go:10 main\nruntime/internal/atomic/types.go:194                    (*Uint32).Load\nruntime/asm_amd64.s:1695                                goexit"

**Reply by james-d-elliott** [reply]
You've put a private key in the `certificates_directory`, it should only contain X509 certs.

**Reply by Livefour2day** [reply]
Thank you James you nailed it!!!!
Sorry for being stupid still learning but love Authelia, thank you for your time with this.
